### DISCUSSION TITLE
Lint only modified files

### DISCUSSION THREAD

**Discussion by gc** [biomejs/biome#4621] [Ideas]
Suggestion: A way to instruct biome to lint only files that have been changed according to git.
My current solution is doing this: `git diff --name-only --diff-filter=d HEAD | xargs biome check --write`. (This command uses git to get a list of all file names of changed files, excluding deleted files, and passes them to biome)

Reason: A full project checkwrite takes 2 seconds in my project, and it's one of the things our developers run to prepare their code to push to github, so this 2 seconds is a painfully long time. Linting only changed files takes this from 2 seconds to being instant in almost all cases. CI tests still lint the entire project, so there's no chance anything was missed. A simple option to make biome run much faster seems like a clear win.

I would suggest a flag/setting that tells biome to only process files changed according to git, and could be used when checking or writing. If, it's accepted as a feature idea, I would consider trying to PR it myself.

**Note:**
There exists a --changed flag already in biome, it does something different and can't be used to do this.
There exists a `--staged` flag, but it effects only _staged_ files, the suggested flag would lint _all_ changes.

**Comment by dyc3** [comment]
I think you are mistaken. Unless I'm missing something, it sounds like `--changed` already does what you want it to do. From the docs:

> add the `--changed` option to your command, to process only those files that your VCS acknowledged as “changed”

Source: https://biomejs.dev/guides/integrate-in-vcs/#process-only-changed-files

**Reply by gc** [reply]
Try it, it doesn't work like that. my understanding is `--changed` is meant to be only used in CI to check things changed by a PR. I have found other people in the discord server getting confused by this, I think its not only me. After looking more, I found `biome check --staged` https://github.com/biomejs/biome/issues/2296, but it only checks staged files. So there is no option to just "lint changed files". 

The 2 flags seem confusing or inconsistently named, I'm not suggesting they be renamed as I just want a way to lint changed files, but I feel like this would make more sense:
`--staged`: checks staged files
`--changed`: checks changed files 
`--changed --since=X` (use this for the current behavior of --changed)

**Reply by ematipico** [reply]
`--staged` is exactly the option you're looking for. If it doesn't work, it could be a bug. The reason why it was called staged, is because that's how git calls the process of adding new files to a repository, mainly via `git add`.

 Feel free to file a bug with a minimal reproduction with instructions.

**Reply by gc** [reply]
If you're saying that `--staged` is meant to lint _unstaged_ changes, that isn't how it appears to work (and isn't what the docs say). If you're not saying that, it's definitely not the option I'm looking for or what I've suggested here.

**Reply by ematipico** [reply]
It wasn't clear to me you were suggesting to include also files that aren't staged. "Changed" was a very broad term :)

You might want to specify this in the suggestion

**Reply by gc** [reply]
All staged files are changed, not all changed files are staged, so this option would lint staged files too. I will edit my suggestion to make it clear.

**Comment by jpambrun** [comment]
I also assumed `--changed` would include "modified" files, or it would be nice to provide a `--modified` flag.

**Comment by hornta** [comment]
I also ran into this just now. I assumed the argument linted all the files that were modified in the repository, but it looks to be the diff between two branches. 

Can we have a another argument that lints these files(which also respects files.include)
<img width="55" height="88" alt="image" src="https://github.com/user-attachments/assets/1543d96e-21c5-4bca-b84f-2e9a3eb626ec" />
